### PR TITLE
scli: pass python dependencies directly

### PR DESCRIPTION
**Pull request by ahldr**
###### Motivation for this change
This is the pattern recommended by @jonringer when I added the package, was changed by @dotlambda in a recent update PR and I think it should stay in line with Jon's recommendations.

See:
- https://github.com/NixOS/nixpkgs/pull/127797
- https://github.com/NixOS/nixpkgs/pull/122315

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dotlambda** [comment]
If @jonringer really advised this pattern, I'd like to know his reasoning. I strongly advise against it because it prevents the use of `packageOverrides`.

**Comment by dotlambda** [comment]
> Read the original PR for some more context: #122315

I can't find any comment by @jonringer in that thread.

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -26610,7 +26610,7 @@ in
 
   scite = callPackage ../applications/editors/scite { };
 
-  scli = callPackage ../applications/misc/scli { };
+  scli = with python3Packages; callPackage ../applications/misc/scli { };
```
Comment (dotlambda) [inline pkgs/top-level/all-packages.nix:26613]:
```suggestion
  scli = python3Packages.callPackage ../applications/misc/scli { };
```

**Comment by alexeyre** [comment]
my bad it was in this nimbo PR https://github.com/NixOS/nixpkgs/pull/121633, but as a fellow python package I think it's still relevant

**Review by dotlambda — CHANGES_REQUESTED** [review]

**Comment by alexeyre** [comment]
https://github.com/NixOS/nixpkgs/pull/121633#discussion_r625426595

> Using pythonXPackages isn't allowed from python-modules as this will mix interpreter versions for a given package. For example, if you were to do python37Packages.<mypkg>, but the expression pulled from python3[8]Packages. Python will be mis-importing several libraries, and this will likely cause issues at runtime.
> Please refer to Python dependency handling on the best practices for handling dependencies within python-modules. Namely, you should be importing each python package individually, and not refer to another python package set (E.g. python3Packages).

**Comment by dotlambda** [comment]
> Namely, you should be importing each python package individually

That's true - for the case of a python module.

**Comment by alexeyre** [comment]
I'm not personally attached to this pattern or anything, but I spent quite a lot of time getting told I was doing it wrong and this was the setup people seemed to be happy with, if you think it's superior I'll trust you

**Comment by dotlambda** [comment]
> I spent quite a lot of time getting told I was doing it wrong

You were, because you were trying to use `python3Packages` in a Python module. But then you turned it into an application. Now it's not wrong but advised against.

**Comment by alexeyre** [comment]
Sorry maybe I came off badly, I know I was doing it wrongly. I think I've mixed patterns here and I obviously didn't catch that the direct-import style is only advised for if it's a python module.

Thanks for showing me this it was pretty likely I would have done this on other packages later down the road if I hadn't been shown this, I appreciate it :)

**Comment by dotlambda** [comment]
In e.g. https://github.com/NixOS/nixpkgs/blob/1f76c234b9942bff08e009bbe46114e2ee327ee3/pkgs/servers/mautrix-telegram/default.nix, you can see that in order to use a specific version of a Python dependency the argument should be `python3`. To be prepared for this eventuality it's best to use the pattern right away.

**Pull request closed without merging**